nav2_costmap_2d/plugins/static_layer.cpp

```diff
@@ -249,9 +251,15 @@ void
 StaticLayer::incomingMap(const nav_msgs::msg::OccupancyGrid::SharedPtr new_map)
 {
   std::lock_guard<Costmap2D::mutex_t> guard(*getMutex());
-  processMap(*new_map);
   if (!map_received_) {
     map_received_ = true;
+    processMap(*new_map);
+  }
+  if (processing_layer_.load()) {
+    map_buffer_ = new_map;
+  } else {
+    processMap(*new_map);
+    map_buffer_ = nullptr;
```

Comment:
You don't need to set to nullptr here. If you set it, it will be reset below in the update function.

Reply:
I think it is necessary. If the incomingMap is called while updating the layer, and then again after the update is finished, the last call should pe processed, and the buffer emptied. Otherwise in this case the new map would be processed twice, once here, another time in the update function.

Reply:
If map_buffer_ is set to nullptr to begin with (you should do that in constructor's `:`), then it is only not nullptr when 259 sets it to something else. In the next function, you process the map and then reset it to `nullptr`. At that point, I don't think there's any case when this is not nullptr where setting it here is necessary. You have a mutex locked above this so only this callback or one of the functions can be in use at the same time, effectively locking this variable -- unless I'm missing something.

Reply:
Yeah, so both functions can call the processMap, therefore if we don't clear the map_buffer_ here, when we get to the update we see that efectively there is something in the buffer and we update it again. I think. The fact of having the lock doesn't change that: When we are here, this function has the lock, and processes the map. Then the update is waiting, and when the lock gets free, it would check the buffer, and call processMap again if it's not empty. I think you are seeing the practical case, but in theory, it would be possible to call incomingMap several times between calls to updateBounds, including during the execution of updateBounds.

But I see how this can be confusing, and tricky to follow. Maybe it's best to simply get the layered_costmap mutex at the beginning of processMap and leave the rest as it was.. 

I'll initialize the buffer to nullptr also, I missed that one.

Reply:
I guess let me game this out without the 262 `map_buffer_ = nullptr`:

- I start up (so map_buffer_ is nullptr) have a new map, not in the middle of a process cycle so it just directly updates. Do this 10 times, still never set to anything other than nullptr 
- I start up (so map_buffer_ is nullptr) have a new map, not in the middle of a process cycle so it just directly updates. Then I update bounds 10 times, so still nullptr so never gets called in the update function
- I start up (so map_buffer_ is nullptr) have a new map, its now in the middle of an update cycle so it stores the map in `map_buffer_` between the 2 functions after the mutex is released and then in the updateCosts it uses the old map since there's no processMap() call in updateCosts. Then we wait for the next iteration of the costmap to add / process the map
- I start up (so map_buffer_ is nullptr) have a new map, its now in the middle of an update cycle so it stores the map in `map_buffer_` between the 2 functions after the mutex is released and then in the updateCosts it uses the old map since there's no processMap() call in updateCosts. Then we have a second new map before the updateBounds can be called again. If we're not in an update cycle, then that new map will be directly processed but the old map will still be stored and processed next round.

You are correct, we need to set to reset in case there's something there. Thanks!

Reply:
Just make the other changes and we should be GTG

Reply:
Okay I'll push the update during the day. 
Just to clarify why I placed the processMap at the beginning of updateBounds and not at the end of updateCosts: 
The line [168 of layered_costmap.cpp](https://github.com/ros-planning/navigation2/blob/d7494ec8c501522200089c54947232fe86c0ef94/nav2_costmap_2d/src/layered_costmap.cpp#L168) can prevent the updateCosts call from being done, which would lock the costmap if the reason for the error comes from the static_layer's map (if that would be the case, the logic would prevent the static map from being updated). By placing the processMap call at the beginning of updateBounds, we ensure that the map would be updated every time there is a new map, without any real changes on the performance of the layered costmap.

Reply:
Okay, I'll fix the branch and the other changes and push during the day.

Just to clarify why I placed the call to processMap at the beginning of updateBounds and not at the end of updateCosts:

The line [168 of layered_costmap.cpp](https://github.com/ros-planning/navigation2/blob/d7494ec8c501522200089c54947232fe86c0ef94/nav2_costmap_2d/src/layered_costmap.cpp#L168) can prevent the call of updateCosts. This is problematic when the reason why we end up returning there is related to the static layer, because in that case, the incomingMap call would be prevented from updating the map (the 'update_in_progress' flag is set to true until the end of updateCosts). 
It might seem like a lot of conditions should be met for that to happen, but it could happen. By moving the processMap to the beginning of updateBounds, we prevent this logic lock from happening without any performance penalty.

Reply:
Beginning makes sense, no issue